### ISSUE TITLE
Symbols/cirlces box-shadows?

### ISSUE THREAD

**Issue by Strongleong** [maplibre/maplibre-gl-js#1326]
Hello! I am wandering will be there ability to make shadows icons/text/circles/etc? I can make circle shadow by draw 2 circles where one of them with blur and opacity, but it is kinda suck.
And I can't make it work with symbol icon (sdf). I am try this with halo-blur and halo-opacity, but it just do not work out. 

P.S. Sorry for bad English :sweat_smile:

**Comment by drwestco** [comment]
The icon-halo-* properties should be applicable here, but looks like they don't work as expected. With a circular SDF icon, the halo is either always a single pixel border, or fills the entire bounds of the icon.

icon-halo-width: 5
![image](https://user-images.githubusercontent.com/743030/179048949-509249e6-2a5e-4a77-bb5c-dcd3b407855f.png)

icon-halo-width: 6
![image](https://user-images.githubusercontent.com/743030/179049018-d1f2a1aa-8d32-4646-bd80-4b5c172a81a8.png)

When we want to add shadows to icons, we bake them directly into the images. This also lets us have offset drop shadows, where the halo remains centered around the icon.

**Comment by Strongleong** [comment]
Yes, I meant exactly that problem with sdf. And I can,t bake shadow to image because sdf restricted to only one color at a time

**Comment by drwestco** [comment]
Yeah, it only works for full color images. Having a set of icon-shadow properties on the style layer would be a nice feature.

**Comment by Strongleong** [comment]
And about `icon-halo`/`circle-stroke`. It is not good solution, because I am currently want shadow **and** halo/stroke.

**Comment by mxzinke** [comment]
> icon-halo-width: 5
> ![](https://user-images.githubusercontent.com/743030/179048949-509249e6-2a5e-4a77-bb5c-dcd3b407855f.png)
>
> icon-halo-width: 6
> ![](https://user-images.githubusercontent.com/743030/179049018-d1f2a1aa-8d32-4646-bd80-4b5c172a81a8.png)

The issue with the `icon-halo-width` still persists in v5 at SDF icons. When using a width bigger than the (relative) padding between icon and origin icon image, the entire bounds of the icons are filled.

_Quick solution is to create SDF spritesheet with bigger paddings at each of the icon._ A more robust solution would be enhancing the current rendering to go over the actual bounds (of the icon image), or at least extending the bounds.

@HarelM I know that you know where I can find the appropriate code. Can you point me to it?

**Comment by HarelM** [comment]
Some of the things related to halo are here in the symbol program that runs in webgl:
https://github.com/maplibre/maplibre-gl-js/blob/cd170349d30ceb4548ba011698814850b5b39bae/src/render/program/symbol_program.ts#L51
Also the draw symbol that sends stuff to the program:
https://github.com/maplibre/maplibre-gl-js/blob/cd170349d30ceb4548ba011698814850b5b39bae/src/render/draw_symbol.ts#L405

**Comment by mxzinke** [comment]
![Image](https://github.com/user-attachments/assets/dcfb6e79-6bca-41d1-9c8a-880af29af3c3)

I am really not into the whole codebase and need to give up after several hours. As you can see on the right, the round halo is cut of the borders. Plus, some translucent background is added around it (maybe some wrong alpha calculations in shaders?). This happens with basically on every SDF image.

Here is the visual test `style.json` I created:

```json
{
    "version": 8,
    "metadata": {
        "test": {
            "width": 128,
            "height": 64,
            "description": "Tests that SDF icons with different halo widths render correctly when approaching padding limits"
        }
    },
    "center": [0, 0],
    "zoom": 0,
    "sources": {
        "geojson": {
            "type": "geojson",
            "data": {
                "type": "FeatureCollection",
                "features": [
                    {
                        "type": "Feature",
                        "properties": {
                            "halo_width": 3
                        },
                        "geometry": {
                            "type": "Point",
                            "coordinates": [-20, 0]
                        }
                    },
                    {
                        "type": "Feature",
                        "properties": {
                            "halo_width": 6
                        },
                        "geometry": {
                            "type": "Point",
                            "coordinates": [20, 0]
                        }
                    }
                ]
            }
        }
    },
    "sprite": "local://sprites/sprite",
    "layers": [
        {
            "id": "background",
            "type": "background",
            "paint": {
                "background-color": "lightblue"
            }
        },
        {
            "id": "symbol",
            "type": "symbol",
            "source": "geojson",
            "layout": {
                "icon-image": "dot.sdf"
            },
            "paint": {
                "icon-color": "blue",
                "icon-halo-color": "white",
                "icon-halo-width": ["get", "halo_width"]
            }
        }
    ]
}
```
(Background can be removed later, just for more visibility)

To fix it, I tried to scale up the whole buffer so that the complete halo fits in. But I failed to find the right spot for implementation of such a solution.

Maybe scaling up can be done by calculating a bigger shape?

https://github.com/maplibre/maplibre-gl-js/blob/cd170349d30ceb4548ba011698814850b5b39bae/src/symbol/shaping.ts#L936C1-L949C2

Are I getting the right thoughts on that? Is there any documentation on how the icon size and so is calculated.

**Comment by HarelM** [comment]
I don't think there's much here, the repo has some readme and developers guide but not much more.
Maybe @xabbu42 know something as he was messing around with this code a while back...
